src/reconcile/rec_col.c

```diff
@@ -277,36 +275,51 @@ __wt_rec_col_int(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *pageref)
-         * replaced, the page's modify structure references it and we built the value cell just
-         * above in the switch statement. Else, the WT_REF->addr reference points to an on-page cell
-         * or an off-page WT_ADDR structure: if it's an on-page cell and we copy it from the page,
-         * else build a new cell.
+         * replaced, the page's modify structure references it and we assigned it just above in the
+         * switch statement. Otherwise, the WT_REF->addr reference points to either an on-page cell
+         * or an off-page WT_ADDR structure: if it's an on-page cell we copy it from the page, else
+         * build a new cell.
          */
         if (addr == NULL && __wt_off_page(page, ref->addr))
             addr = ref->addr;
-        if (addr == NULL) {
+        if (addr != NULL) {
+            __wt_rec_cell_build_addr(session, r, addr, NULL, ref->ref_recno, page_del);
+            WT_TIME_AGGREGATE_COPY(&ta, &addr->ta);
+        } else {
             __wt_cell_unpack_addr(session, page->dsk, ref->addr, vpack);
-            if (F_ISSET(vpack, WT_CELL_UNPACK_TIME_WINDOW_CLEARED)) {
-                /* Need to rebuild the cell with the updated time info. */
-                __wt_rec_cell_build_addr(session, r, NULL, vpack, ref->ref_recno, NULL);
+            if (cms.state == WT_CHILD_PROXY || F_ISSET(vpack, WT_CELL_UNPACK_TIME_WINDOW_CLEARED)) {
+                /*
+                 * Need to build a proxy (page-deleted) cell or rebuild the cell with updated time
+                 * info. If we don't have page-delete information already, propagate any that exists
+                 * in the cell. This can be needed if we get back WT_CHILD_ORIGINAL and the time
+                 * window gets cleared.
+                 */
+                if (page_del == NULL && vpack->type == WT_CELL_ADDR_DEL)
+                    page_del = &vpack->page_del;
+                __wt_rec_cell_build_addr(session, r, NULL, vpack, ref->ref_recno, page_del);
             } else {
+                /* Copy the entire existing cell, including any page-delete information. */
                 val->buf.data = ref->addr;
                 val->buf.size = __wt_cell_total_len(vpack);
                 val->cell_len = 0;
                 val->len = val->buf.size;
             }
             WT_TIME_AGGREGATE_COPY(&ta, &vpack->ta);
-        } else {
-            __wt_rec_cell_build_addr(session, r, addr, NULL, ref->ref_recno, NULL);
-            WT_TIME_AGGREGATE_COPY(&ta, &addr->ta);
+        }
+        if (page_del != NULL) {
+            ft_ta.newest_start_durable_ts = ta.newest_start_durable_ts;
+            ft_ta.newest_stop_durable_ts = page_del->durable_timestamp;
+            ft_ta.oldest_start_ts = ta.oldest_start_ts;
+            ft_ta.newest_txn = page_del->txnid;
+            ft_ta.newest_stop_ts = page_del->timestamp;
+            ft_ta.newest_stop_txn = page_del->txnid;
```

Comment:
It feels like there should be macros to help with this - do they not work because the time window is spread across the page_del and another time aggregate?

Reply:
Indeed.

There's now two copies of this though (one here and one in rec_row) so maybe it should get its own macro.

Reply:
Fixed in [c296405](https://github.com/wiredtiger/wiredtiger/pull/8129/commits/c296405e0efafc45182121a1d390c1a1adfb69d0).

I tidied up the way it works some also.